custom_components/spotcast/helpers.py

```diff
@@ -103,117 +107,324 @@ async def run(*args, loop=None, executor=None, **kwargs):
+        pass
+
+    results = spotify_client.artist_top_tracks(artistUri)
+    for track in results["tracks"][:10]:
+        _LOGGER.debug("track    : " + track["name"])
+
+    return results["tracks"]
+
+
+def get_search_string(
+    artistName: str,
+    albumName: str,
+    trackName: str,
+    genreName: str,
+    playlistName: str,
+    showName: str,
+    episodeName: str,
+    audiobookName: str,
+) -> str:
+    search = []
+    if not is_empty_str(artistName):
+        search.append(f"artist:{artistName}")
+        search.append(artistName)
+    if not is_empty_str(albumName):
+        search.append(f"album:{albumName}")
+        search.append(albumName)
+    if not is_empty_str(trackName):
+        search.append(f"track:{trackName}")
+        search.append(trackName)
+    if not is_empty_str(genreName):
+        search.append(f"genre:{genreName}")
+        search.append(genreName)
+    # if we are searching for a playlist, podcast, audiobook, we need some search query which is probably just the text we are looking for
+    for item in [playlistName, showName, episodeName, audiobookName]:
+        if not is_empty_str(item):
+            search.append(item)
+
+    return " ".join(search)
+
+
+# "album", "artist", "playlist", "track", "show", "episode", "audiobook"
+def get_types_string(
+    artistName: str,
+    albumName: str,
+    trackName: str,
+    playlistName: str,
+    showName: str,
+    episodeName: str,
+    audiobookName: str,
+) -> str:
+    types = []
+    if not is_empty_str(artistName):
+        types.append("artist")
+    if not is_empty_str(albumName):
+        types.append("album")
+    if not is_empty_str(trackName):
+        types.append("track")
+    if not is_empty_str(playlistName):
+        types.append("playlist")
+    if not is_empty_str(showName):
+        types.append("show")
+    if not is_empty_str(episodeName):
+        types.append("episode")
+    if not is_empty_str(audiobookName):
+        types.append("audiobook")
+    return ",".join(types)
+
+
+def get_search_results(
+    spotify_client: spotipy.Spotify,
+    limit: int = 10,
+    country: str = None,
+    artistName: str = None,
+    albumName: str = None,
+    playlistName: str = None,
+    trackName: str = None,
+    showName: str = None,
+    episodeName: str = None,
+    audiobookName: str = None,
+    genreName: str = None,
+):
+    _LOGGER.debug("using search query to find uri")
+    searchResults = []
+
+    if (
+        not is_empty_str(artistName)
+        and len(
+            list(
+                filter(
+                    lambda x: not is_empty_str(x),
+                    [
+                        albumName,
+                        playlistName,
+                        trackName,
+                        showName,
+                        episodeName,
+                        audiobookName,
+                        genreName,
+                    ],
+                )
             )
+        )
+        == 0
+    ):
```

Comment:
this condition should succeed only if we are searching for an artist and nothing else, no other fields. In that case, do top tracks. Any other case, don't do top tracks, for example, "artist: the beatles" + "genre: electronic", we don't just want top tracks by the beatles, we want their electronic tracks (Tomorrow Never Knows?)